### PR TITLE
perf(lang): precompile ohm grammars to recipes

### LINKED ISSUE #9025: @usebruno/lang compiles all six ohm grammars at module load, costing ~100ms on every require
### Summary

`@usebruno/lang` compiles all six of its ohm PEG grammars **from source at module load**, so
`require('@usebruno/lang')` costs ~100ms before anything is parsed. The CLI pays it on every
invocation, and so does every consumer of `@usebruno/filestore`, which imports lang at its top level.

### Details

The grammars are built by top-level `const`s, so the cost is paid on `require`, not on first parse:

```
packages/bruno-lang/v2/src/bruToJson.js:43                   const grammar = ohm.grammar(`Bru {
packages/bruno-lang/v2/src/envToJson.js:24                   const grammar = ohm.grammar(`Bru {
packages/bruno-lang/v2/src/collectionBruToJson.js:16         const grammar = ohm.grammar(`Bru {
packages/bruno-lang/v2/src/example/bruToJson.js:16           const exampleGrammar = ohm.grammar(`Example {
packages/bruno-lang/v2/src/example/request/bruToJson.js:19   const requestGrammar = ohm.grammar(`Request {
packages/bruno-lang/v2/src/example/response/bruToJson.js:13  const responseGrammar = ohm.grammar(`Response {
```

`rollup.config.js` does not list `ohm-js` in `externalDeps`, which reads as though the grammars get
inlined at build time — they do not. This is grammar *evaluation*, not module loading, so bundling and
tree-shaking cannot remove it; profiles put the time in `ohm-js/src/pexprs-eval.js`, `MatchState.js`
and `Semantics.js`.

### Measurements

M-series Mac, Node 22. Compile cost per grammar, measured in a fresh process each time:

| grammar | compile |
| --- | --- |
| `Bru` (request) | ~26ms |
| `Request` (example) | ~21ms |
| `Response` (example) | ~16ms |
| `Bru` (collection) | ~13ms |
| `Example` | ~10ms |
| `Bru` (environment) | ~7ms |
| **total `ohm.grammar()`** | **~93ms** |

`require('@usebruno/lang')` is ~102ms end to end, of which `require('ohm-js')` itself is only ~10ms.

Reproduce:

```bash
node -e "const t=performance.now();require('@usebruno/lang');console.log((performance.now()-t).toFixed(1)+'ms')"
```

Note the three `example/*` grammars are easy to under-measure: `bruToJson` requires
`example/bruToJson`, so timing them after it reports ~0ms. Measure each in a fresh process.

### Who pays it

- **bruno-cli** — every invocation, on top of its other startup costs.
- **Any `@usebruno/filestore` consumer** — `filestore/src/index.ts:23` imports `@usebruno/lang` at the
  top level, so importing filestore costs ~135ms, most of it grammar compilation.
- **bruno-electron** — once at boot, via the IPC handlers and watchers. Negligible against multi-second
  app startup, so not worth optimising for on its own.

It matters most for anything invoking the CLI repeatedly — CI pipelines, and tools that spawn
`bru run` per request.

### Suggested fix

ohm supports serializing a compiled grammar to a "recipe" (`grammar.toRecipe()` /
`ohm.makeRecipe()`) that rebuilds in well under a millisecond. Pre-generating recipes removes the cost
rather than deferring it, which lazy construction would not — the grammars are needed as soon as
anything is parsed.

Whatever validates a recipe needs to be strict enough that a stale one falls back to compiling from
source rather than being used: it should be tied to the ohm version that produced it and to the
grammar source it describes, so an edited grammar or an ohm upgrade degrades to slow-but-correct.

### PR DESCRIPTION
**Pull request by W-Floyd**
### Description

`@usebruno/lang` compiles all six of its ohm grammars from source at module load, so
`require('@usebruno/lang')` costs ~100ms before anything is parsed. This pre-generates each grammar as
an ohm "recipe" and rebuilds from that instead, falling back to compiling from source whenever the
recipe no longer matches.

`require('@usebruno/lang')` drops from **~102ms to ~24ms**.

Closes #9025

### Problem

The grammars are top-level `const`s, so the cost lands on `require`, not on first parse. `~93ms` of the
`~102ms` is `ohm.grammar()` itself; `require('ohm-js')` is only ~10ms.

`rollup.config.js` does not list `ohm-js` in `externalDeps`, which reads as though the grammars are
inlined at build time. They are not — this is grammar *evaluation*, so bundling and tree-shaking cannot
remove it.

Paid by the CLI on every invocation, and by every `@usebruno/filestore` consumer, since
`filestore/src/index.ts` imports lang at its top level.

### Fix

`compileGrammar(key, source)` prefers a pre-generated recipe from `generated/<key>.json`. The
six grammar modules change by two lines each; **no grammar rule text is touched**.

A recipe is used only when it still describes the grammar in front of it — it must come from this
ohm-js version, record the hash of this source, and embed this source verbatim:

- **ohm version** matters because `Builder.fromRecipe` only throws on operations it does not
  recognise, so a compatible-looking recipe from another compiler could rebuild into something subtly
  different instead of falling back. `ohm-js` is a floating `^16.6.0`, so a minor bump is a real path.
- **Embedded source**, not just the hash beside it, catches a recipe payload swapped between two files
  by a bad merge. The recorded grammar name would not — three of these grammars are called `Bru`.
- **Hashing ignores line endings**, so a CRLF checkout keeps its recipes instead of silently paying
  full price. (`.gitattributes` pins `eol=lf`, so this only affects clones predating it or a tool that
  rewrites a file, but the failure was invisible.)

Anything else — an edited grammar, an ohm upgrade, a corrupt or mismatched file — falls back to
compiling from source. **A stale recipe is only ever slower, never wrong.**

### The recipes are build output, not source

They are git-ignored, written to `packages/bruno-lang/generated` by the package's `prepare` step, and
published via `files`. `npm run generate:grammars` regenerates them by hand (aliased at the repo root,
like the other `build:*` steps).

`prepare` is what makes this work in both directions, and I verified both rather than assuming:

- it runs during `npm install` for a workspace package, **after** that package's dependencies are
  linked — so `ohm-js` is available to the generator, and a plain install produces the recipes for the
  `bruno-cli` and `bruno-electron` builds, which use the workspace copy rather than a tarball
- it also runs before `npm pack`/`publish`, and `files` carries the output into the tarball even though
  git ignores it — all six recipes are packed, and `scripts/` still is not

An earlier revision of this PR committed the recipes, on the reasoning that generation could only
happen at publish time and would therefore leave the CLI and electron builds still paying the ~90ms.
That was wrong: a `prepare` step covers both, so there was no trade-off to make.

### Verification

- **Parsing is identical.** All 724 `.bru` fixtures in the repo through the four v2 entry points —
  **2896 parse attempts, 0 diffs**, comparing both the JSON output and the thrown ohm error message
  strings (full diagnostics with line/col and `Expected …` sets), so source-interval fidelity survives
  the round-trip.
- **The recipe path is the one actually taken** — instrumenting `ohm.makeRecipe`/`ohm.grammar` shows
  `recipeCalls=6, sourceCalls=0`. Without this check the equivalence comparison above could have been
  two source compiles agreeing with each other.
- **Every fallback verified to fire**, each still parsing correctly: wrong ohm version, wrong source
  hash, swapped recipe payload, missing payload, unreadable JSON, and no recipe at all. Each is now a
  test asserting the grammar was compiled *from source* — spying on `ohm.makeRecipe` — rather than
  merely that parsing still works, and each was checked to fail when its guard is removed.
- `npm test --workspace=packages/bruno-lang` — **475 passing** (33 suites); `--workspace=packages/bruno-filestore`
  — **202 passing**.
- **The build step is covered too**, not just described: the suite asserts every registered grammar has
  a recipe file on disk, that the directory they are actually written to is one `files` publishes, and
  that running the generator exits cleanly leaving every recipe matching its source — so a broken
  `prepare` fails in CI rather than at publish time.
- Drift is enforced in CI: the spec checks every *registered* grammar has a fresh recipe, so a new
  grammar cannot quietly ship without one, and an edited grammar fails with the regenerate command.
- Removing `v2/src/generated/` entirely leaves parsing correct and the suite green apart from the
  recipe-specific specs — the fallback is not theoretical.

### Notes for reviewers

- The recipes no longer appear in the diff at all, being git-ignored build output. They are single-line
  and ~50KB each, so they were never human-reviewable anyway. What guards them is
  the spec's full structural comparison — `makeRecipe(recipe).toRecipe()` against the source-compiled
  grammar's `toRecipe()` — which fails on any divergence. Rebuilding a recipe does not execute code:
  `makeRecipe` walks the array via `Builder.fromRecipe`, with no `eval` or `new Function`.
- `bruno-lang` stays a leaf with zero internal `@usebruno/*` dependencies. The new `fs`/`crypto` use is
  Node built-ins in a Node-side package; `bruno-app` does not import lang, and `bruno-filestore` lists
  it as rollup-external, so the recipe read never reaches a browser bundle.

### Screenshots

Not applicable — no UI change.

#### Contribution Checklist:

- [x] **I've used AI significantly to create this pull request**
- [x] **The pull request only addresses one issue or adds one feature.**
- [x] **The pull request does not introduce any breaking changes**
- [ ] **I have added screenshots or gifs to help explain the change if applicable.** — not applicable
- [x] **I have read the [contribution guidelines](https://github.com/usebruno/bruno/blob/main/contributing.md).**
- [x] **Create an issue and link to the pull request.** — #9025
- [x] **I've run the claude code review skill locally.**

## Summary by CodeRabbit

* **New Features**
  * Added automated grammar generation and packaging for more reliable grammar availability.
  * Added grammar caching to improve initialization efficiency.
  * Added fallback handling when cached grammar data is missing or invalid.

* **Bug Fixes**
  * Improved handling of outdated or mismatched grammar data.
  * Preserved consistent parsing across different line-ending formats.

* **Tests**
  * Expanded coverage for grammar generation, validation, caching, recovery, and packaging.

### CODE REVIEW

File: packages/bruno-lang/v2/tests/grammar-cache.spec.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      expect(grammar.match('a').succeeded()).toBe(true);
+    });
+
+    it('hashes line endings alike, so a CRLF checkout still matches its recipe', () => {
+      const [key] = keys;
+      const { source } = grammars.get(key);
+
+      expect(grammarCache.hashSource(source.replace(/\n/g, '\r\n'))).toBe(grammarCache.hashSource(source));
+    });
+  });
```
[the remark this answers is not in the document]
Comment (W-Floyd) [reply packages/bruno-lang/v2/tests/grammar-cache.spec.js:174]: Fair hit — fixed in 39cabbe.

Both of those paths existed and worked, but you are right that nothing in CI proved it. I had verified them by hand while writing the guards and then deleted the throwaway probe, which is exactly the gap: the behaviour was checked once and left unguarded.

Three cases added, each asserting the grammar was compiled *from source* rather than merely that parsing still succeeds — the spy on `ohm.makeRecipe` is what distinguishes a real fallback from a recipe that happened to work:

- a recipe built by a different ohm-js is ignored (`makeRecipe` never called)
- a recipe whose metadata all checks out but which ohm cannot rebuild falls back after `makeRecipe` throws — `['notAnOhmOperation', { source }]` passes the version, hash and embedded-source checks, so it exercises the `try/catch` specifically
- a recipe that is not readable as JSON is ignored

The guards read from disk, so each writes a throwaway recipe file and removes it in a `finally`, with an `afterAll` asserting nothing is left beside the real recipes.

I also checked the tests are not vacuous, since a guard test that passes with the guard removed is worse than none: deleting the `ohmVersion` comparison fails the first, and deleting the `try/catch` around `makeRecipe` fails the second. 468 lang and 202 filestore tests pass.

**Comment by sid-bruno** [comment]
Hey, thanks for the PR. 

I don't think the generated files need to be a part of the code and can be added into the `.gitignore`, we can add in a build step and correct the directories that are published with the code. If you have the time, it'd be great if that was done. 

Another reason for why the caching was avoided is that the project itself is moving towards `.yml` instead of `.bru` altogether so we've reduced our efforts on making sure bru gets faster (replace ohm-js / upgrade to 18 for wasm , etc) 

If the 100ms drop for a collection is making a difference for you, then we'll see if we can get the PR in

**Comment by W-Floyd** [comment]
To clarify, it matters more for me when running bru-cli when it pays the 100ms cost on every invocation. Also, it costs the 100ms even if using `.yaml` files, so this helps all cases until `.bru` support is fully removed (if ever).

File: packages/bruno-lang/v2/tests/grammar-cache.spec.js [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    it('regenerates every recipe by running the generator', () => {
+      // `prepare` runs this script on install and before publish. If it were broken, the recipes would
+      // quietly stop being produced and the startup cost would come back, so run it for real.
+      const generator = spawnSync(process.execPath, [path.join(PACKAGE_ROOT, 'scripts/generate-grammars.js')], {
+        cwd: PACKAGE_ROOT,
+        encoding: 'utf8'
+      });
+
+      expect(generator.stderr).toBe('');
+      expect(generator.status).toBe(0);
```
[the remark this answers is not in the document]
Comment (W-Floyd) [reply packages/bruno-lang/v2/tests/grammar-cache.spec.js]: Agreed, and fixed in 174d2b5.

Empty stderr was never something the generator promised — a clean run that happened to warn would have failed the test for no reason, which is the wrong thing to pin. The exit status is the contract, so that is what is asserted now, and stderr goes into the failure message where it is actually useful:

```js
if (generator.status !== 0) {
  throw new Error(`Generator exited with ${generator.status}:\n${generator.stderr}`);
}
```

I checked the distinction holds rather than assuming it, since the point of the change is that one case should start passing while the other keeps failing: a generator that writes to stderr and exits 0 now passes, and one that exits non-zero still fails, reporting `Generator exited with 1:` followed by its output.

File: packages/bruno-lang/v2/tests/grammar-cache.spec.js [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+      for (const key of keys) {
+        const recipe = grammarCache.readRecipe(key);
+        expect(recipe.sourceHash).toBe(grammarCache.hashSource(grammars.get(key).source));
```
[the remark this answers is not in the document]
Comment (W-Floyd) [reply packages/bruno-lang/v2/tests/grammar-cache.spec.js:81]: Fixed in 174d2b5.

You are right that this was the worst possible failure mode for a diagnostic test: `readRecipe` returns `null` for anything missing or unreadable, so dereferencing it turned a real generator bug into a `TypeError` that named no key — leaving whoever hit it to work out which of six grammars broke.

It now fails with the key:

```js
const recipe = grammarCache.readRecipe(key);
if (!recipe) {
  throw new Error(`Generator left no readable recipe for "${key}"`);
}
```

475 lang tests pass, lint clean.
